raylib: init at 3.5.0

adamlwgriffiths opened the pull request:
###### Motivation for this change

Add raylib ~~3.0.0~~ 3.5.0
Fixes #98029

Execution was tested using the provided examples for both static (rlgl_standalone) and dynamic (other examples).

TODO:
* Optionally supports Wayland, wasn't sure how to handle that in nix packages properly
* Supports Android, again, unsure how to handle that.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

adamlwgriffiths commented:
Opened up for review to seek help for the audio issue.
Also any advice on the wayland / android packaging would be good. Otherwise I think it's ok to just merge with X11 support for now.

OPNA2608 reviewed (changes requested):
Tested example `audio_module_playing`. A fatal `SIGSEGV` aside, `strace` shows attempts to open `libpulse` and `libasound`, though these aren't declared dependencies of either the raylib ELF library nor the compiled binary. This will need abit more wrapping / ELF wrangling to work.

![grafik](https://user-images.githubusercontent.com/23431373/93228634-ff9dfe00-f775-11ea-9f35-822b78c350b2.png)

Additionally the pkgconfig file in `lib/pkgconfig/raylib.pc` has broken paths and will need some more attention as well.

```pc
prefix=/nix/store/kxlh5s1p5ld9bxk22wrl1zpglx493apw-raylib-3.0.0
exec_prefix=${prefix}
libdir=${exec_prefix}//nix/store/kxlh5s1p5ld9bxk22wrl1zpglx493apw-raylib-3.0.0/lib
includedir=${prefix}//nix/store/kxlh5s1p5ld9bxk22wrl1zpglx493apw-raylib-3.0.0/include
```

adamlwgriffiths commented:
Thanks for the helpful comments. I'll address them tomorrow =)

adamlwgriffiths committed:
maintainers: Add adamlwgriffiths

adamlwgriffiths commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,32 @@
+{ stdenv, fetchurl, cmake, alsaLib, libX11, libXi, libXcursor, libXrandr, libXinerama, mesa, libGLU, glfw }:
```
Ok will update.

adamlwgriffiths commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,32 @@
+  pname = "raylib";
+  version = "3.0.0";
+
+  src = fetchurl {
+    url = "https://github.com/raysan5/raylib/archive/3.0.0.tar.gz";
+    sha256 = "164d1cc1710bb8e711a495e84cc585681b30098948d67d482e11dc37d2054eab";
+  };
```
Will update

adamlwgriffiths commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,32 @@
+    sha256 = "164d1cc1710bb8e711a495e84cc585681b30098948d67d482e11dc37d2054eab";
+  };
+
+  buildInputs = [ cmake alsaLib libX11 libXi libXcursor libXrandr libXinerama mesa libGLU glfw ];
```
I wasn't aware of the split. Thanks, will update.

adamlwgriffiths replied:
Raylib supports wayland windowing.
I haven't got any experience with wayland. If wayland windowing were enabled, would it still require libX*?
Or does `[ wayland wayland-protocols ]` cover that?

adamlwgriffiths commented on maintainers/maintainer-list.nix:
```diff
@@ -158,6 +158,12 @@
     githubId = 124545;
     name = "Anthony Cowley";
   };
+  adamgriffiths = {
+    email = "adam.lw.griffiths@gmail.com";
+    github = "adamlwgriffiths";
+    githubId = 1239156;
+    name = "Adam Griffiths";
+  };
```
Will do.
This will require a force push. It seems github handles that ok from what I've seen.

adamlwgriffiths commented:
> Tested example `audio_module_playing`. A fatal `SIGSEGV` aside, `strace` shows attempts to open `libpulse` and `libasound`, though these aren't declared dependencies of either the raylib ELF library nor the compiled binary. This will need abit more wrapping / ELF wrangling to work.
> 
> ![grafik](https://user-images.githubusercontent.com/23431373/93228634-ff9dfe00-f775-11ea-9f35-822b78c350b2.png)
> 
> Additionally the pkgconfig file in `lib/pkgconfig/raylib.pc` has broken paths and will need some more attention as well.
> 
> ```
> prefix=/nix/store/kxlh5s1p5ld9bxk22wrl1zpglx493apw-raylib-3.0.0
> exec_prefix=${prefix}
> libdir=${exec_prefix}//nix/store/kxlh5s1p5ld9bxk22wrl1zpglx493apw-raylib-3.0.0/lib
> includedir=${prefix}//nix/store/kxlh5s1p5ld9bxk22wrl1zpglx493apw-raylib-3.0.0/include
> ```

I don't fully grok this one.
I've got alsaLib listed as a buildInput, I would've thought that would provide it at run-time.
I'll keep looking through the docs to see what I've misunderstood.

I'll look into the raylib.pc, strange that it's done that.

adamlwgriffiths committed:
raylib: init at 3.5.0

adamlwgriffiths commented:
I've pushed the trivial changes you mentioned above.
Still todo (help would be great):
* audio not working
* pkgconfig/raylib.pc paths incorrect

I'd also like to expose the wayland option (Cmake flag USE_WAYLAND=ON).
Using [this as a reference](https://github.com/NixOS/nixpkgs/blob/86afe1e4d70eba4df12d0799486c920560903a93/pkgs/applications/misc/bemenu/default.nix) it's not too hard, I've got the option working locally. I don't however have wayland to test with.
Do you know if, when using wayland, you still require the libX* libraries as dependencies? Or does `[ wayland wayland-protocols ]` cover that?

adamlwgriffiths commented:
Ok, took a further look and the USE_WAYLAND is just fed into GLFW, which makes sense. So I'll remove that option. Looks like the x11 libs are required regardless then.

adamlwgriffiths commented:
If I add
```
LD_LIBRARY_PATH="${alsaLib}/lib";
```
to the definition, then audio plays - at least in the nix-shell used to build the library. But I have a feeling that's not the correct solution here.

IvarWithoutBones commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,39 @@
+  # disable -Werror=format-security until new version (>3.0.0) is released with this fix
+  # https://github.com/raysan5/raylib/pull/1305
+  hardeningDisable = [ "format" ];
+
```
Should fix the audio problem:
```suggestion
  postFixup = ''
    wrapProgram $out/lib/*.so --prefix LD_LIBRARY_PATH : {alsaLib}/lib
  '';
```

adamlwgriffiths replied:
Ah, I was looking at this but wasn't sure if it worked for shared objects or just executables.
Fantastic, I'll try this out tomorrow!

adamlwgriffiths replied:
Ok, so I can get this to work when patching the example binaries, but not the .so files themselves.
```
  postFixup = ''
    wrapProgram $out/../source/build/examples/audio_music_stream --prefix LD_LIBRARY_PATH : ${alsaLib}/lib
  '';
```

If I patch the .so files I get:
```
$ ./audio_music_stream 
./audio_music_stream: error while loading shared libraries: <snip>/out/lib/libraylib.so.301: invalid ELF header
```

OPNA2608 replied:
`wrapProgram` only wraps executables afaik, not dynamic libraries.

> If I patch the .so files I get:
> `[…] invalid ELF header`

If you `cat` that "library", it most likely placed a script there that sets `LD_LIBRARY_PATH` then tries to `exec` the wrapped library. That'll definitely not work for this.

OPNA2608 commented:
First things first, you prolly want to make `alsaLib` and `libpulseaudio` optional inputs to support non-Linux systems and Linux systems that, for example, don't want PulseAudio installed in any way. Enabling them by default should be fine though. E.g.

```nix
{ stdenv
# …
, alsaSupport ? stdenv.hostPlatform.isLinux
, alsaLib
, pulseSupport ? stdenv.hostPlatform.isLinux
, libpulseaudio
# …
}:

stdenv.mkDerivation rec {
  # …
  buildInputs = [
    # …
  ]
  ++ lib.optional alsaSupport alsaLib
  ++ lib.optional pulseSupport libpulseaudio;
  # …
}
```

---

> I've got alsaLib listed as a buildInput, I would've thought that would provide it at run-time.
> I'll keep looking through the docs to see what I've misunderstood.

> I'm still confused as to why this is occuring to begin with. Surely Nix is aware it needs alsa, I've given it to the build process, but it's not in the LD_LIBRARY_PATH.

A dependency in `buildInputs` will be installed at runtime, there is no promise though that code will always be able to find it. Nix automagically fixes declared ELF dependencies on libraries during compilation - you can view these for example by running `ldd` on a binary / library. These are *always required*, and the dynamic loader will immediately abort if any of those listed dependencies cannot be found.

![grafik](https://user-images.githubusercontent.com/23431373/93299995-4d5a4b00-f7f6-11ea-9c66-530a48b62d23.png)

But that's not the only way a library can be loaded, and it's not always the preferred method. Presumably in the case of raylib (assuming both ALSA and PulseAudio are used at buildtime for their headers), the audio dependencies are just *"enabled"*, not *"required"*. This means that the code *could* use PulseAudio if it's available at runtime, but if it cannot be found then execution can continue and it may, for example, gracefully fall back to trying ALSA next. Such mechanisms require additional investigation and wrapping to function.

There are multiple ways to resolve this. Generally, polluting `LD_LIBRARY_PATH` can lead to more complicated problems down the line w/r/t dynamically loading other libraries / programs further down the line. I think you could use `patchelf` or another mechanism to add `alsaLib` and `libpulseaudio` to `raylib`'s rpath if support for them has been enabled, then those libraries should be discoverable by the dynamic linker.

```nix
{
postFixup = if (stdenv.hostPlatform.isLinux && (alsaSupport || pulseSupport)) then ''
  rpath=$(patchelf --print-rpath $out/lib/libraylib.so)
  patchelf --set-rpath "$rpath:${lib.makeLibraryPath (lib.optionals alsaSupport [ alsaLib ] ++ lib.optionals pulseSupport [ libpulseaudio ])}" $out/lib/libraylib.so
'' else ""; # or a section for stdenv.hostPlatform.isDarwin w/ install_name_tool if needed
}
```

This can be made to look prettier ofc and I didn't test this, stuck bisecting another PR's regressions rn sorry. Hope this more detailed explanation helps you understand the gist of the problem!

adamlwgriffiths commented:
No that's a very helpful response and makes sense.
I appreciate both of your time. This turned out to be a rougher introduction to nix packages than I had expected.
Again, I'll have another go with your suggestions tomorrow.
If I face any more issues I'll post on the IRC forums.

adamlwgriffiths commented:
Great, that worked.
I'm just supporting Linux at the moment as I don't have easy access to the other targets. I may look into a BSD VM given time.

The only remaining issue is the pkconfig/raylib.pc file.

OPNA2608 commented:
> ```pc
> prefix=@CMAKE_INSTALL_PREFIX@
> exec_prefix=${prefix}
> libdir=${exec_prefix}/@CMAKE_INSTALL_LIBDIR@
> includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@
> ```

You should notify upstream that they construct those paths in their pkg-config file improperly. You can link them to [jtojnar/cmake-snips](https://github.com/jtojnar/cmake-snips) for a better explanation, specifically the sections "**Assuming `CMAKE_INSTALL_<dir>` is relative path**" and "**Concatenating paths when building pkg-config files**".

If you can't wait until upstream fixes this though, then there are patches in Nixpkgs you could adapt to it in the meantime, see [pkgs/development/libraries/yder/fix-pkgconfig.patch](https://github.com/NixOS/nixpkgs/blob/601221807c7d78aaf7407824f20e0b9a52dec733/pkgs/development/libraries/yder/fix-pkgconfig.patch). They're *also* improper but should fix this specific problem for now, just make sure to add a comment linking it to the upstream issue so it can be removed once a proper fix is in place and it's time to update this package. :slightly_smiling_face:

adamlwgriffiths commented:
* Updated to Raylib 3.5.0.
* Removed CMake patch to fix building as it was merged upstream.
* Remove the "hardeningDisable" as this is fixed upstream.

AndersonTorres commented:
We need tests!

adamlwgriffiths commented:
Ok, I'll have to look into how to do that.
There is an issue with 3.5.0 where the examples don't compile.
I fix was applied post-3.5.0. I can apply this as a patch if needed.
Instead I disabled the example programs being built (they would be included with the lib which is unnecessary).

It seems however, the examples are the closest thing to tests that raylib has. In that they are not.
Being a graphics library, testing those tends to be quite difficult (I think game studios do image deltas).
I'm not sure if there's anything I can do here, short of adding my own tests to raylib?

AndersonTorres commented:
I really don't know a way to do this. Usually I test less graphical things, just text diffs.
But images are other beast. A pixel difference is way less important than one character difference.

AndersonTorres commented:
Just to know: the original tests contained in the upstream project aren't running, but the lib works nonetheless?

adamlwgriffiths commented:
Sorry I didn't get notification of your subsequent messages =/

The lib works for me locally. I use it as part of a nix-shell environment and have used both nim and lua bindings with it.

There are no actual unit tests. It seems the closest thing to a test is the examples.

The examples in 3.5.0 have a bug that prevents them being used to demonstrate the library.
[This PR](https://github.com/raysan5/raylib/pull/1470) fixes it, but there hasn't been a release that includes it.
I doubt there's any value in including the patch as the examples shouldn't be shipped with the lib.

If you want to test, you can apply the patch and re-enable compiling the examples by removing ["-DBUILD_EXAMPLES=OFF" on line 30](https://github.com/NixOS/nixpkgs/pull/98030/files#diff-838e9b7462f672bffed6da564f8bc74eac907daf6eaf7c4b3720c1f7117dc056R30).

AndersonTorres commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,46 @@
+    "-DUSE_EXTERNAL_GLFW=ON"
+    "-DSHARED=ON"
+    "-DBUILD_EXAMPLES=OFF"
+  ] ++ lib.optional includeEverything "-DINCLUDE_EVERYTHING=ON";
```
What does `includeEverything` do?

adamlwgriffiths replied:
Enables all the optional image codecs.

https://github.com/raysan5/raylib/wiki/CMake-Build-Options

Binary Option | Description | Default Value
-- | -- | --
INCLUDE_EVERYTHING | Include all optional off-by-default formats in build | OFF
SUPPORT_FILEFORMAT_PNG | Support loading PNG as textures | ON
SUPPORT_FILEFORMAT_DDS | Support loading DDS as textures | ON
SUPPORT_FILEFORMAT_HDR | Support loading HDR as textures | ON
SUPPORT_FILEFORMAT_KTX | Support loading KTX as textures | ON
SUPPORT_FILEFORMAT_ASTC | Support loading ASTC as textures | ON
SUPPORT_FILEFORMAT_BMP | Support loading BMP as textures | OFF
SUPPORT_FILEFORMAT_TGA | Support loading TGA as textures | OFF
SUPPORT_FILEFORMAT_JPG | Support loading JPG as textures | OFF
SUPPORT_FILEFORMAT_GIF | Support loading GIF as textures | OFF
SUPPORT_FILEFORMAT_PSD | Support loading PSD as textures | OFF
SUPPORT_FILEFORMAT_PKM | Support loading PKM as textures | OFF
SUPPORT_FILEFORMAT_PVR | Support loading PVR as textures | OFF
SUPPORT_FILEFORMAT_OBJ | Support loading OBJ file format | ON
SUPPORT_FILEFORMAT_MTL | Support loading MTL file format | ON

AndersonTorres replied:
They are auto-detected in the `cmakeConfigurePhase`?

adamlwgriffiths replied:
These defines enable support for writing to stdout the various image file formats. Relevant code is here: [stb_image_write.h](https://github.com/raysan5/raylib/blob/82d61a5875c4a7eaa4c8caf9ed3b19a8441eb258/src/external/stb_image_write.h).

[stb_image.h](https://github.com/raysan5/raylib/blob/82d61a5875c4a7eaa4c8caf9ed3b19a8441eb258/src/external/stb_image.h) handles the reading of various images and always supports those file formats.

There are no extra dependencies to enable this functionality.

AndersonTorres commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,46 @@
+    ${lib.optionalString pulseSupport "patchelf --add-needed ${libpulseaudio}/lib/libpulse.so $out/lib/libraylib.so.${version}"}
+  '';
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

AndersonTorres commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,46 @@
+  '';
+
+  meta = with stdenv.lib; {
+    description = "raylib is a simple and easy-to-use library to enjoy videogames programming";
```
```suggestion
    description = "A simple and easy-to-use library to enjoy videogames programming";
```

`description` should not include the `pname` in it.

adamlwgriffiths commented:
Thanks for the helpful comments!
I'll look at getting the patch to fix the examples in tonight.

AndersonTorres commented:
@adamlwgriffiths please squash the suggestions into the original commit, maintaining only two (the one adding you to the maintainers and the other creating the expression).

adamlwgriffiths commented:
I know you edited your comment to remove the request to add the patch, but I went a long anyway (why else do anything eh?).
I'm having trouble though, and I can't seem to break through what's wrong.

I've added the following section
```
  patches = [
    (fetchpatch {
      # fixes examples not compiling in 3.5.0
      url = "https://patch-diff.githubusercontent.com/raw/raysan5/raylib/pull/1470.patch";
      sha256 = "be4d519ec2ac93bb6e7b18a50447a0900dc105182496deec69c3a2469b43ad53";
    })
  ];
```

But now I get the following error:

```hash mismatch in fixed-output derivation '/nix/store/c2vnjrkabkz2wasqzr7nfsf0k9z2a0nm-1470.patch':
  wanted: sha256:0lxd8fdld8n3d7ndx5i4302w23chl13h998qgdpbp4xcqag52kdy
  got:    sha256:1ff5l839wl8dxwrs2bwky7kqa8kk9qmsflg31sk5vbil68dzbzg0
error: build of '/nix/store/d93f74mbshikx2chdx3zffnfhyr5ni9p-source.drv', '/nix/store/n9l2mk70cfl030b08dylb94483y9jxy7-1470.patch.drv' failed
```

It seems thats the resulting hash of the patched source and derivation file, as changing the derivation also changes the hash.
But I can't find any examples in nixpkgs or any documentation online that talks about this being a problem.
The hash for raylib hasn't changed, and the hash for the patch should be valid.
```
$ wget https://patch-diff.githubusercontent.com/raw/raysan5/raylib/pull/1470.patch
$ sha256sum 1470.patch 
be4d519ec2ac93bb6e7b18a50447a0900dc105182496deec69c3a2469b43ad53  1470.patch
```

Setting the patch's hash to another value still causes the same issue. I get a different "wanted" hash, but the same "got" hash.
Confused.

AndersonTorres commented:
@adamlwgriffiths, ah, this is a confusing history.

1. fetchpatch is not identical to fetchurl. The fetchpatch function reads the hash and strips the irrelevant information contained in it. Only after that stripping it calculates the hash.

    This is similar to removing the comments of a Python source code: the program runs the exact same way, but the hash of the source code changes.

2. In Nixpkgs, we use the infamous *Trust On First Use* protocol. We put a fake hash sum and do a first try. The program will complain about a wrong hash and will provide you the "got" hash. Then you copy that got hash in the place of the new one.

adamlwgriffiths commented:
That's where I'm confused.
The "wanted" doesn't match the sha I have listed, infact it's neither the src nor patch hashes.
If I set the patch hash to the "got" value nothing changes, it still lists the previous wanted and got hashes.
It seems the "wanted" is a derived hash?

Perhaps its how I'm trying to test it.
I'm using the following:
`$ eval 'nix-shell -E "with import <nixpkgs> {}; callPackage /home/adam/Workspace/nixpkgs/pkgs/development/libraries/raylib/default.nix {}"'`

AndersonTorres commented:
Show me all the output, please.

Is the ` <nixpkgs>`  the most recent one?

adamlwgriffiths commented:
```
$ cat shell.sh 
set -e
eval 'nix-shell -E "with import <nixpkgs> {}; callPackage /home/adam/Workspace/nixpkgs/pkgs/development/libraries/raylib/default.nix {}"'

$ ./shell.sh 
these derivations will be built:
  /nix/store/d93f74mbshikx2chdx3zffnfhyr5ni9p-source.drv
  /nix/store/n9l2mk70cfl030b08dylb94483y9jxy7-1470.patch.drv
building '/nix/store/n9l2mk70cfl030b08dylb94483y9jxy7-1470.patch.drv'...
building '/nix/store/d93f74mbshikx2chdx3zffnfhyr5ni9p-source.drv'...

trying https://patch-diff.githubusercontent.com/raw/raysan5/raylib/pull/1470.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

trying https://github.com/raysan5/raylib/archive/3.5.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   121  100   121    0     0    199      0 --:--:-- --:--:-- --:--:--   199
100  3626    0  3626    0     0   4893      0 --:--:-- --:--:-- --:--:--  4893
hash mismatch in fixed-output derivation '/nix/store/c2vnjrkabkz2wasqzr7nfsf0k9z2a0nm-1470.patch':
  wanted: sha256:0lxd8fdld8n3d7ndx5i4302w23chl13h998qgdpbp4xcqag52kdy
  got:    sha256:1ff5l839wl8dxwrs2bwky7kqa8kk9qmsflg31sk5vbil68dzbzg0
error: build of '/nix/store/d93f74mbshikx2chdx3zffnfhyr5ni9p-source.drv', '/nix/store/n9l2mk70cfl030b08dylb94483y9jxy7-1470.patch.drv' failed
```

No, nixpkgs is the version from this PR.

AndersonTorres commented:
Try to use this `sha256 = 1ff5l839wl8dxwrs2bwky7kqa8kk9qmsflg31sk5vbil68dzbzg0;` in fetchPatch

adamlwgriffiths commented:
omg sorry, its just one of those days. I was editing the same file but in a different directory.
*facepalm*

AndersonTorres commented:
Oh, it *haspens*.

AndersonTorres commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,54 @@
+  };
+
+  patches = [
+    (fetchpatch {
+      # fixes examples not compiling in 3.5.0
```
```suggestion
    # fixes examples not compiling in 3.5.0
    (fetchpatch {
```

AndersonTorres reviewed (changes requested):
A final change and everything will be OK.

adamlwgriffiths commented:
Applied and squashed.
Testing locally in a nix-shell and the examples are working well.

adamlwgriffiths commented on pkgs/development/libraries/raylib/default.nix:
```diff
@@ -0,0 +1,54 @@
+  cmakeFlags = [
+    "-DUSE_EXTERNAL_GLFW=ON"
+    "-DSHARED=ON"
+    "-DBUILD_EXAMPLES=OFF"
```
It looks like the examples dont end up in the final output directory, only remaining in the build dir.
Is there any value in re-enabling them?

AndersonTorres replied:
Well, let me know if I understood it correctly.

The examples generate executable programs and they work, but these generated executable files are not pushed in the final dir?

If this is the case, there is no problem. After all, these examples are merely for testing purposes, and there is no serious reason to save the results of a test. (Except maybe if some fun and useful thing comes with them, like the famous `glxgears`.) 

On the other hand, it is useful to save the source code of these examples. Put these codes in a directory like `$out/share/doc/${pname}/examples/`

adamlwgriffiths replied:
Thanks for the suggestion. I appreciate you taking the time to hand-hold me through this.

Looking at the docs, the best phase I could find was the `postInstall` phase.
I copy all the example source to $out/share/doc/${pname}/examples
Then cleanup the root directory which has some CMakeLists.txt, Makefile, *.bat, and other files which shouldn't be relevant.

I looked through the nixpkgs repo and couldn't find any other libraries that build nor copy the examples (SDL/2, cegui, mygui, etc).
(As an aside, I also was unable to get github search to work. I have a feeling the repo is too big)

I think it's good to copy the examples, as you say.
I'm thinking it might be good to build and copy the example binaries too.
Thoughts?
Would this be in `$out/share/bin`?

The only issue is, you would double up on all the assets, as they are in the example files next to the relevant source.
Then CMake moves them to the collective location with the output binaries.
Perhaps this is just more trouble than it's worth and what we've got now is sufficient?

AndersonTorres replied:
Yes, the Nixpkgs is a huge monster nowadays. It has almost 2G (and `pkgs/` has 134M) according to `du -sh`.
But you can search the local copy with an utility like ripgrep.

About the installation: Well, I am thinking a bit about it.
The example binaries are not so useful. The end users can copy the source codes of these examples and compile them by themselves.
If the CMake scripts of Raylib does not install these example binaries, then leave them out - forget them.

My suggestion is: 

- Don't install the source code of the examples - yes, forget them;
- If the CMake scripts don't install the examples binaries, forget them too.

Yes, it is much squeeze for little juice.

adamlwgriffiths replied:
Sorry, when you say "enable the tests", do you mean build the examples but not package them?

AndersonTorres replied:
Yes. Use the examples in a `checkPhase`, but don't install them.

adamlwgriffiths replied:
The examples are graphical and require the user to press ESC to exit, they aren't simple CLI programs which quit.
I wouldn't be able to execute them, which I think is what you're recommending I do.
I think this is where our cross-communication is coming in?

I can re-enable them as part of the build, but I think there would be little value there if they're not executed.

AndersonTorres replied:
Well, the mere fact they are compiling without errors is a successful test, even if they are interactive.

But you are right, there is little value in these tests, at least in our current framework. Disable the examples.

(NixOS has a unit test framework, based on virtual machines, and the interactive programs can be controlled on it. But forget this for now.)

adamlwgriffiths commented:
Ok, I think that covers everything then.
Thanks for your help Anderson, much appreciated!
